service/security/shiro/src/main/java/org/eclipse/kapua/service/authentication/token/shiro/AccessTokenServiceImpl.java

```diff
@@ -226,9 +226,13 @@ public void invalidate(KapuaId scopeId, KapuaId accessTokenId) throws KapuaExcep
         // Do find
         entityManagerSession.onTransactedResult(em -> {
             AccessToken accessToken = AccessTokenDAO.find(em, scopeId, accessTokenId);
-            accessToken.setInvalidatedOn(new Date());
+            if (accessToken != null) {
+                accessToken.setInvalidatedOn(new Date());
 
-            return AccessTokenDAO.update(em, accessToken);
+                return AccessTokenDAO.update(em, accessToken);
+            } else {
+                return null;
```

Comment:
Hi @muros-ct ,

why not throw KapuaNotFoundException?

Regards,

_Alberto_

Reply:
Instead of returning null you mean. Like:

`throw new KapuaEntityNotFoundException(AccessToken.TYPE, scopeId);`

The problem might be that AccessTokenServiceImpl.invalidate() is called from logout, and if that one receives exception it doesn't complete the logout and next time user tries to login it is denied, as it is effectively still logged in. After first failed attempt it can than login successfully, as it is logged out by unsuccessful login. Complicated I know :)

Reply:
I would throw the `KapuaEntityNotFoundException` and catch it in the caller:

AuthenticationServiceShiroImpl.logout();
AuthenticationServiceShiroImpl.refreshAccessToken();

and manage in there the exception.\

Would it make more sense? 

Regards, 

_Alberto_

Reply:
Sure, I will give it a try. I think it should work.

Reply:
@Coduz I commited changes.